src/Lachain.Core/Vault/KeyGenManager.cs

```diff
@@ -64,10 +64,15 @@ private void BlockManagerOnSystemContractInvoked(object _, InvocationContext con
         {
             if (context.Receipt is null) return;
             var highestBlock = _blockSynchronizer.GetHighestBlock();
-            var willParticipate =
+            var willParticipate = 
                 !highestBlock.HasValue ||
-                GovernanceContract.IsKeygenBlock(highestBlock.Value) &&
+                GovernanceContract.IsKeygenBlock(context.Receipt.Block) &&
```

Comment:
Take a look at this change: I don't sure I understand this condition good enough, but I think it is more correct than before (and it is really works for small cycles,  I've checked it with debugging exactly this place)